test/integration/certificates/trigger_controller_test.go

```diff
@@ -96,3 +97,109 @@ func TestTriggerController(t *testing.T) {
+	// as we want to test the very specific case of triggering due to a renewal being
+	// required
+	policyChain := policies.Chain{policies.CurrentCertificateNearingExpiry(fakeClock)}
+	// Build, instantiate and run the trigger controller.
+	_, factory, cmCl, cmFactory := framework.NewClients(t, config)
+	ctrl, queue, mustSync := trigger.NewController(logf.Log, cmCl, factory, cmFactory, framework.NewEventRecorder(t), fakeClock, policyChain)
+	c := controllerpkg.NewController(
+		logf.NewContext(context.Background(), logf.Log, "trigger_controller_RenewNearExpiry"),
+		"trigger_test",
+		metrics.New(logf.Log),
+		ctrl.ProcessItem,
+		mustSync,
+		nil,
+		queue,
+	)
+	stopController := framework.StartInformersAndController(t, factory, cmFactory, c)
+	defer stopController()
+
+	// Create a Certificate resource and wait for it to have the 'Issuing' condition.
+	cert, err := cmCl.CertmanagerV1alpha2().Certificates("testns").Create(ctx, &cmapi.Certificate{
+		ObjectMeta: metav1.ObjectMeta{Name: "testcrt", Namespace: "testns"},
+		Spec: cmapi.CertificateSpec{
+			SecretName: "example",
+			CommonName: "example.com",
+			IssuerRef:  cmmeta.ObjectReference{Name: "testissuer"}, // doesn't need to exist
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Ensure that the Certificate does *not* get the Triggered status condition
+	// if the status.renewalTime is not set.
+	// Wait for 2s, polling every 200ms to ensure that the controller does not set
+	// the condition.
+	ensureCertificateDoesNotHaveIssuingCondition(ctx, t, cmCl, cert.Namespace, cert.Name)
+
+	t.Logf("Setting status.renewalTime in the future on Certificate resource")
+	renewalTime := metav1.NewTime(fakeClock.Now().Add(time.Second))
+	cert.Status.RenewalTime = &renewalTime
+	cert, err = cmCl.CertmanagerV1alpha2().Certificates(cert.Namespace).UpdateStatus(ctx, cert, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Log("Ensuring Certificate does not have Issuing condition for 2s...")
+	ensureCertificateDoesNotHaveIssuingCondition(ctx, t, cmCl, cert.Namespace, cert.Name)
+
+	t.Log("Advancing the clock forward to renewal time")
+	// advance the clock to a millisecond after the renewal time, as the
+	// fakeclock implementation uses .After when checking whether to
+	// trigger timers.
+	fakeClock.SetTime(renewalTime.Time.Add(time.Millisecond))
+
+	err = wait.Poll(time.Millisecond*200, time.Second*2, func() (done bool, err error) {
+		c, err := cmCl.CertmanagerV1alpha2().Certificates(cert.Namespace).Get(ctx, cert.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if apiutil.CertificateHasCondition(c, cmapi.CertificateCondition{
+			Type:   cmapi.CertificateConditionIssuing,
+			Status: cmmeta.ConditionTrue,
+		}) {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Error("Failed waiting for Certificate to have Issuing condition")
+	}
+}
```

Comment:
I think this all makes sense. It's a shame to have to wait 2s for something *not* to happen, and I wondered if it might be neater to also run the readiness controller here with a shared fake clock and watch that the readiness controller first updates the Status.RenewalTime field, without any change to the Isssuing condition, and then advance the clock beyond the renewal time and then wait for the Issuing condition. Might be more trouble than it's worth, but maybe create a followup issue to add tests for the controllers that update and use the new RenewalTime  fields.

Reply:
Yeah, I'm definitely interested to see if we can write some more all-encompassing integration tests that instantiate all controllers at once, although this then veers closer and closer towards being an end-to-end test really.

Right now the other integration tests we have test a single `expcertificates` controller at a time, and it at least allows us to test specific behaviour of the controllers in a quite easy-to-read way.

Having to wait 2s is definitely a shame - the only way I can think to avoid this is through a `status.observedGeneration` field.